tests/test_factory.py

```diff
@@ -1060,3 +1060,247 @@ def test_poetry_build_system_dependencies(
     poetry = Factory().create_poetry(temporary_directory)
 
     assert set(poetry.build_system_dependencies) == expected
+
+
+def test_create_poetry_with_nested_dependency_groups(temporary_directory: Path) -> None:
+    pyproject_toml = temporary_directory / "pyproject.toml"
+    content = """\
+[project]
+name = "my-package"
+version = "1.2.3"
+
+[tool.poetry.group.testing.dependencies]
+pytest = "*"
+pytest-cov ="*"
+
+[tool.poetry.group.dev]
+include-groups = [
+    "testing",
+]
+[tool.poetry.group.dev.dependencies]
+black = "*"
+"""
+    pyproject_toml.write_text(content)
+    poetry = Factory().create_poetry(temporary_directory)
+
+    assert len(poetry.package.all_requires) == 5
+    assert [
+        (dep.name, ",".join(dep.groups)) for dep in poetry.package.all_requires
+    ] == [
+        ("pytest", "testing"),
+        ("pytest-cov", "testing"),
+        ("black", "dev"),
+        ("pytest", "dev"),
+        ("pytest-cov", "dev"),
+    ]
+
+
+def test_create_poetry_with_self_referenced_dependency_groups(
```

Comment:
`parametrize` with different names (test normalization).

Reply:
I find it hard to parametrize those test, without hiding what is actually tested to much. I've introduce a helper method to avoid some code duplication. Now I find one could follow the purpose of the tests, there input and the expected outcome quite good.

Reply:
As long as at least one test fails if I remove (exactly) one arbitrary `canonicalize_name`, I am fine.

Reply:
Ah, now I guess I get what you want to parametrize. :sweat_smile: I added some.